### PR TITLE
Feature probe level corner

### LINKED ISSUE #549: [FR] Calling the M48 command from the menu
Hello,
When calling M48, the message - first homing XYZ is displayed.
Solution:
add another line with the G28 command to the "leveling.c" file before the M48 command G28.

For example:
case KEY_ICON_4:
   storeCmd ("G28 \ n");
   storeCmd ("M48 \ n");
break;

M48

### LINKED ISSUE #1102: [FR] (standard deviation on repeatability test )
**Is your feature request related to a problem? Please describe.**

it would be nice to also be able to enter the standard deviation along with the other data when the test repeatability is performed.

### LINKED ISSUE #1646: [FR] Use probe to level corners
Possible to use probe to level corners in Marlin Mode when #define LEVEL_CORNERS_USE_PROBE is enabled in Marlin but there is no option to do this in Touch Mode. It would be nice if it could be implemented.

**Description**
Add a option to level the bed corners using a probe. Instead of using the nozzle to paper ajust the screws, it will probe the first corner (or the center) and save the z position. Then will go to the next corner to check if is whitin tolerance range. If is in tolerance, will go to next corner. If is not, will leave the probe deployed waiting the user to raise the bed util the probe is trigered. When all corners are within tolerance, will present a option to execute bed leveling.

https://github.com/MarlinFirmware/Marlin/pull/20241

### PR DESCRIPTION
**Pull request by Cal1sto**
### Requirements

New Feature request to have the Marlin function of Level Corner on the TFT with more function and for best leveling experience with a probe.

### Description

The Level corner function added in ABL menu give the potentiality to quick level the bed by measuring with the probe the 4 corner and adjust it, the tested point could be adjusted like the manual leveling

### Benefits

Easy and very quick bed leveling could be performed with this function and usefull for bed level check.

### Function

Level corner (L corner):  give you the following function , it will probe all the 4 corners , display the Z value measured on each corner, then you could adjust each corner individualy and you could use your own tolerance based on the probe precision that you could get with the M48 . You could scan all corner again and again or check and adjust individualy the 4 corners in cross adjustement or linear adjustement manually. Like the manual leveling function you have also the possibility to set the corner position test, this is limited already with a minimum value set depending of the probe offset to the nozzle to reduce the risk the printing head is not going outside the X and Y bed area, set a reasonable lowest value to measure nearest the 4 corners leveling screw of the bed (default value will be set automatically on the lowest possible offset based on the probe offset to be accurate and optimal).

### Implements / Resolves

resolves #1646
resolves #1102
resolves #549

### CODE REVIEW

**Comment by oldman4U** [comment]
Nice👍🏻

**Review by digant73 — CHANGES_REQUESTED** [review]

File: TFT/src/User/API/parseACK.c [hunk 1]
```diff
@@ -677,18 +712,27 @@ void parseACK(void)
...
         {
           setParameter(P_PROBE_OFFSET, Z_STEPPER, ack_value());
         }
       }
+      
       // parse G29 (ABL) + M118, ABL Completed message (ABL, BBL, UBL) (e.g. from ABL menu)
       else if (ack_seen("ABL Completed"))
       {
-        ablUpdateStatus(true);
+        mblUpdateStatus(true);
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/parseACK.c:735]: no this is wrong  please let ablUpdateStatus(true);

File: TFT/src/User/API/coordinate.c [hunk 2]
```diff
@@ -74,6 +75,16 @@ float coordinateGetAxisTarget(AXIS axis)
 {
   return targetPosition.axis[axis];
 }
+// Set level corner position the measured Z offset from probe, see in ABL.c menu refreshLevelCornerValue(MENUITEMS levelItems) and value get from parseACK.c
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/coordinate.c:78]: Done

File: TFT/src/User/API/coordinate.c [hunk 3]
```diff
@@ -74,6 +75,16 @@ float coordinateGetAxisTarget(AXIS axis)
...
+void SetLevelCornerPosition(int point,float position)
+{
+    levelCornerPosition[point] = position;
+}
+// Get level corner position the measured Z offset from probe, see in ABL.c menu refreshLevelCornerValue(MENUITEMS levelItems) and value get from parseACK.c
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/coordinate.c:83]: Done

File: TFT/src/User/API/coordinate.h [hunk 4]
```diff
@@ -42,7 +42,10 @@ float coordinateGetAxisActual(AXIS axis);
 void coordinateSetAxisActual(AXIS axis, float position);
 void coordinateQuerySetWait(bool wait);
 void coordinateQuery(void);
-
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/coordinate.h:45]: Done

File: TFT/src/User/API/menu.h [hunk 5]
```diff
@@ -132,11 +132,13 @@ typedef struct
 
  typedef struct
 {
- uint8_t   enabled[LIVEICON_LINES];
- LIVE_DATA lines[LIVEICON_LINES];
+ uint8_t   enabled[LIVEICON_LINES+1];
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/menu.h:135]: Done, thanks

File: TFT/src/User/API/menu.h [hunk 6]
```diff
@@ -132,11 +132,13 @@ typedef struct
 
  typedef struct
 {
- uint8_t   enabled[LIVEICON_LINES];
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/menu.h:135]: Done

File: TFT/src/User/API/parseACK.c [hunk 7]
```diff
@@ -444,6 +444,41 @@ void parseACK(void)
       {
         coordinateSetExtruderActualSteps(ack_value());
       }
+      // G30 feedback to get the 4 corners Z value returned by Marlin for LevelCorner function in ABL.c
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/parseACK.c]: Done

File: TFT/src/User/API/parseACK.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/API/parseACK.c]: wrong correction done during compilation, i roll it back already to  ablUpdateStatus

File: TFT/src/User/Configuration.h [hunk 8]
```diff
@@ -5,6 +5,8 @@
 //===========================================================================
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/Configuration.h]: ok, rolled back to original one from bigreetech.master

File: TFT/src/User/Menu/BedLevelingLayer2.c [hunk 9]
```diff
@@ -46,11 +46,15 @@ void menuBedLevelingLayer2(void)
   {
     if (infoSettings.touchmi_sensor != 0)
     {
+      bedLevelingLayer2Items.items[3].icon = ICON_LEVELCORNER;
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply TFT/src/User/Menu/BedLevelingLayer2.c]: Thanks logical ;-) - done

File: platformio.ini [hunk 10]
```diff
@@ -8,409 +8,381 @@
 ; Please visit documentation for the other options and examples
```
[the remark this answers is not in the document]
Comment (Cal1sto) [reply platformio.ini]: Original Platformio.ini restored

**Review by Cal1sto — COMMENTED** [review]
#define STRING_LEVELCORNER            "Level corner" is too long replaced with "L corner"

**Review by Cal1sto — COMMENTED** [review]
Remove ABL.c and ABL.h from this request no change performed there from original, this was done on previous master state

**Comment by Cal1sto** [comment]
LevelCorner.c and LevelCorner.h replace the old implementation in ABL.c and ABL.h done on the old Master state before change with BedLevelingLayer2.c and BedLevelingLayer2.h  -- > remove from this request both ABL.c and ABL.h

**Comment by oldman4U** [comment]
Hi Cal1sto.

Would it be possible to reduce the number of different icons used for this functionality, or to use Icons which are already available? There are point_1 to point_5 bmp's available today which do the same but only without the sensor.

This would be great, thank you

**Comment by Cal1sto** [comment]
Hi oldman4U,

I understood why (limited free space), i will check if i could move the measured value bellow the icone, and find a way to clear the label area to display the value change.

**Comment by Cal1sto** [comment]
Hi oldman4U,

Only one new icone left the levelcorner.bmp all other replaced with the leveling one "POINT_1" .... 
LevelCorner.c, menu.c and icon_list.inc updated / adapted for the new text location

**Comment by Guilouz** [comment]
I redesigned icons to stay in the same spirit as the others.

Sans titre-1

[Level Corner BMP.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/6168020/Level.Corner.BMP.zip)

**Comment by Cal1sto** [comment]
New icone proposal fron Guilouz added

**Comment by bigtreetech** [comment]
Hello, I'm going to release a new stable version of firmware on this weekend(3/21), so major changes won't be merged into the main branch for the time being. When the new version is released, I will merge this PR immediately.

**Comment by Cal1sto** [comment]
Thanks bigtreetech i will inform the community about this good news , take care and have great time

**Comment by Cal1sto** [comment]
The bmp delivered by Guilouz, was defect and could not be read during the import(update) on the device, i open it and just save it in same location, after that the bmp format was ok.

**Comment by Guilouz** [comment]
> The bmp delivered by Guilouz, was defect and could not be read during the import(update) on the device, i open it and just save it in same location, after that the bmp format was ok.

Very strange, they are exported in bmp 24 bits from photoshop

**Comment by Cal1sto** [comment]
Yes i saw that it's very strange (Je sait j'ai vue cela) , I use GIMP for this type of purpose and he was able to read it, but strange he generate 66 bytes more .

**Comment by Guilouz** [comment]
> Yes i saw that it's very strange (Je sait j'ai vue cela) , I use GIMP for this type of purpose and he was able to read it, but strange he generate 66 bytes more .

I will try your PR tonight, i will see the issue with bmp files, they must have all the same size.

**Comment by Cal1sto** [comment]
I found why i provide 66 bytes more and why you provide 2 byte more at the end of the file i commit now the last fully correct files that is readable, and standard.
image

**Comment by oldman4U** [comment]
@Acenotass @AntoszHUN @bepstein111

One Icon more, sorry.

But like BTT wrote, there is no need to hurry, as all new features will be released next week, once the new firmware is available.

Thank you all!

File: TFT/src/User/API/parseACK.c [hunk 11]
```diff
@@ -677,14 +677,22 @@ void parseACK(void)
...
+        if (ack_seen("Y"))
+        {
+          setParameter(P_PROBE_OFFSET, Y_STEPPER, ack_value());
+        }
         if (ack_seen("Z:") || (ack_seen("Z")))
         {
           setParameter(P_PROBE_OFFSET, Z_STEPPER, ack_value());
         }
-      }
+      }
```
Comment (digant73) [inline TFT/src/User/API/parseACK.c:695]: remove the trailing spaces after "}"

File: TFT/src/User/API/parseACK.c [hunk 12]
```diff
@@ -695,7 +703,41 @@ void parseACK(void)
...
-
+      // G30 feedback to get the 4 corners Z value returned by Marlin for LevelCorner function in ABL.c 
+      else if (ack_seen("Bed X: ")) 			   
+      {
+        float valy = 0;
+        float valx = ack_value();
+        if (ack_seen("Y: ")) valy = ack_value();
+        if ((valx < 100) && (valy < 100))		  
+        {
+          if (ack_seen("Z: ")) {
```
Comment (digant73) [inline TFT/src/User/API/parseACK.c:714]: put all the "{" in a new line

File: TFT/src/User/Configuration.h [hunk 13]
```diff
@@ -344,7 +344,6 @@
 /**
```
Comment (digant73) [inline TFT/src/User/Configuration.h:344]: restore the original file

File: TFT/src/User/Menu/ABL.c [hunk 14]
```diff
@@ -174,4 +174,4 @@ void menuUBLLoad(void)
 {
```
Comment (digant73) [inline TFT/src/User/Menu/ABL.c:174]: restore the original file

File: TFT/src/User/Menu/ABL.h [hunk 15]
```diff
@@ -20,4 +20,4 @@ void menuUBLLoad(void);
 }
```
Comment (digant73) [inline TFT/src/User/Menu/ABL.h:20]: restore the original file

File: platformio.ini [hunk 16]
```diff
@@ -413,4 +413,4 @@ build_flags   = ${stm32f10x.build_flags}
   -DRAM_SIZE=64  ; Available RAM size in kbytes
```
Comment (digant73) [inline platformio.ini:413]: restore the original file

**Review by digant73 — CHANGES_REQUESTED** [review]

**Comment by Cal1sto** [comment]
Files restored from source repository, No change to be performed on Configuration.h, platformio.ini, ABL.c and ABL.h

**Comment by AntoszHUN** [comment]
> @Acenotass @AntoszHUN @bepstein111
> 
> One Icon more, sorry.
> 
> But like BTT wrote, there is no need to hurry, as all new features will be released next week, once the new firmware is available.
> 
> Thank you all!

I do not see. What icon? What is the name?

**Comment by oldman4U** [comment]
Look what Guilouz  wrote and attached yesterday.

**Comment by Cal1sto** [comment]
Please could someone advise how to remove the ABL.c, ABL.h, Configuration.h and platformio.ini from this PR, there is no change on this files, i replace all this from origin master, but still on some strange difference :
- New line (char) at the end of the files
- ABL.c Diff not reflecting the reality ? Or wrong compare result ?

**Comment by digant73** [comment]
> Please could someone advise how to remove the ABL.c, ABL.h, Configuration.h and platformio.ini from this PR, there is no change on this files, i replace all this from origin master, but still on some strange difference :
> 
> * New line (char) at the end of the files
> * ABL.c Diff not reflecting the reality ? Or wrong compare result ?

possibly it's just wrong report from github. forget it at the moment. Eventually I will restore the proper files

File: TFT/src/User/API/coordinate.c [hunk 17]
```diff
@@ -75,6 +76,18 @@ float coordinateGetAxisTarget(AXIS axis)
   return targetPosition.axis[axis];
 }
 
+// Set level corner position the measured Z offset from probe, see in ABL.c menu refreshLevelCornerValue(MENUITEMS levelItems) and value get from parseACK.c
+void SetLevelCornerPosition(int point,float position)
+{
+    levelCornerPosition[point] = position;
```
Comment (digant73) [inline TFT/src/User/API/coordinate.c:82]: 2 spaces for indentation, not 4

File: TFT/src/User/API/coordinate.c [hunk 18]
```diff
@@ -75,6 +76,18 @@ float coordinateGetAxisTarget(AXIS axis)
...
+}
+
+// Get level corner position the measured Z offset from probe, see in ABL.c menu refreshLevelCornerValue(MENUITEMS levelItems) and value get from parseACK.c
+float GetLevelCornerPosition(int point)
+{
+    return levelCornerPosition[point];
```
Comment (digant73) [inline TFT/src/User/API/coordinate.c:88]: 2 spaces for indentation, not 4

File: TFT/src/User/Menu/ABL.c [hunk 19]
```diff
@@ -79,8 +79,10 @@ void ablStart(void)
       storeCmd("G29\n");
```
Comment (digant73) [inline TFT/src/User/Menu/ABL.c:79]: merge ABL.c with the last BTT changes

File: TFT/src/User/Menu/BedLevelingLayer2.c [hunk 20]
```diff
@@ -85,6 +88,10 @@ void menuBedLevelingLayer2(void)
         if (infoMachineSettings.leveling == BL_UBL)
           menuUBLLoad();
         break;
+       
+      case KEY_ICON_3:    
+          infoMenu.menu[++infoMenu.cur] = menuLevelCorner;
```
Comment (digant73) [inline TFT/src/User/Menu/BedLevelingLayer2.c:93]: reduce the indetation so it is at teh same level of "break"

File: TFT/src/User/Menu/LevelCorner.c [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+      case KEY_ICON_7:
+        infoMenu.cur--;
+        break;
+      default:
+        break;
+    }
+    refreshLevelCornerValue(levelcornerItems);    
+    loopProcess();
+  }
+}
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:128]: put a new line after "}"

File: TFT/src/User/Menu/LevelCorner.h [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+extern "C" {
+#endif
+
+void menuLevelCorner(void);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.h:14]: put a new line after #endif"

**Review by digant73 — CHANGES_REQUESTED** [review]

**Comment by digant73** [comment]
@Cal1sto if I'm not wrong your PR implements FR #1646, so please at the end of the PR description report the following line:

resolves #1646

**Comment by Cal1sto** [comment]
@digant73 Thanks yes it seems a request related and matched with the FR #1646, PR description updated.

**Comment by Cal1sto** [comment]
@digant73 Could you check if the ABL.c in this PR is now the origin/master (bigtreetech:master) one, i am not sure if i did it right because i don't see anw difference in the files changed section, and then if not i need your help to do it correctly, thanks

**Comment by Cal1sto** [comment]
@bigtreetech You could merge this PR, the conflict resulted to other merge are now solved. thanks

File: TFT/src/User/API/coordinate.c [hunk 23]
```diff
@@ -69,6 +70,18 @@ void coordinateSetAxisTarget(AXIS axis, float position)
   }
 }
 
+// Set level corner position the measured Z offset from probe, see in ABL.c menu refreshLevelCornerValue(MENUITEMS levelItems) and value get from parseACK.c
```
Comment (digant73) [inline TFT/src/User/API/coordinate.c:73]: move your new two functions after coordinateQuery so you will respect the same order in the .h file

File: TFT/src/User/API/parseACK.c [hunk 24]
```diff
@@ -735,7 +743,45 @@ void parseACK(void)
...
+        }
+        else if ((valx < 100) && (valy > 100))   
+        {
+          if (ack_seen("Z: "))
+          {
+            SetLevelCornerPosition(4,ack_value());
+            SetLevelCornerPosition(0, 4);
+          }
+        }
+      }
```
Comment (digant73) [inline TFT/src/User/API/parseACK.c:784]: put a new line after }

File: TFT/src/User/Menu/ABL.c [hunk 25]
```diff
@@ -79,7 +79,7 @@ void ablStart(void)
...
       break;
   }
-
+
```
Comment (digant73) [inline TFT/src/User/Menu/ABL.c:82]: remove the two spaces, just leave the empty line

File: TFT/src/User/Menu/LevelCorner.c [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+  int valIndex[4] = {4,6,2,0};
+  int valPos;
+  int valPosSub;
+  LIVE_INFO lvIcon;
+  lvIcon.lines[0].pos = ss_val_point;
+  lvIcon.lines[1].pos = ss_val_point;
+  lvIcon.lines[2].pos = ss_val_point;
+  lvIcon.lines[3].pos = ss_val_point;
+
+  if ( (int)GetLevelCornerPosition(0) != 0)
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:45]: remove the space after (

File: TFT/src/User/Menu/LevelCorner.c [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+  {
+    if (((int)GetLevelCornerPosition(0) >= 1) && ((int)GetLevelCornerPosition(0) <= 4) )
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:47]: remove the space before the last )

File: TFT/src/User/Menu/LevelCorner.c [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+  } 
+}
+
+void menuLevelCorner(void)
+{ 
+  KEY_VALUES key_num = KEY_IDLE;
+
+  menuDrawPage(&levelcornerItems);
+
+  /* Init Probe Offset in parseAck to get probe offset X and Y*/
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:65]: convert all the /* comments to // (single line comments is preferred)

File: TFT/src/User/Menu/LevelCorner.c [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+  /* Init Coordinate */
+  if (coordinateIsKnown() == false) 
+  {
+    mustStoreCmd("G28\n");
+  }
+  /* Check min edge limit for the probe with probe offset set in parseACK.c */
+  uint8_t edge_min = MAX(ABS(getParameter((s16)P_PROBE_OFFSET, X_STEPPER)),ABS((s16)getParameter(P_PROBE_OFFSET, Y_STEPPER))) + 1;
+  if (infoSettings.level_edge < edge_min) 
+  {
+    infoSettings.level_edge = ((LEVELING_EDGE_DISTANCE >= edge_min)?LEVELING_EDGE_DISTANCE:edge_min);
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:77]: separate "?" and ":" with a space before and after

File: TFT/src/User/Menu/LevelCorner.c [hunk 30]
```diff
@@ -0,0 +1,128 @@
...
+
+  while (infoMenu.menu[infoMenu.cur] == menuLevelCorner)
+  {
+    key_num = menuKeyGetValue();
+    switch (key_num)
+    {
+      case KEY_ICON_0:
+        ScanLevelCorner(3);
+        break;
+      case KEY_ICON_1:
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:94]: put a new line between each "case" statement

File: TFT/src/User/Menu/LevelCorner.c [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+      case KEY_ICON_2:
+        ScanLevelCorner(2);
+        break;
+      case KEY_ICON_3:
+        {
+          char tempstr[30];         
+          sprintf(tempstr, "%Min:%d | Max:%d", edge_min, LEVELING_EDGE_DISTANCE_MAX);
+          int val = numPadInt((u8 *)tempstr, infoSettings.level_edge, LEVELING_EDGE_DISTANCE_DEFAULT, false);
+          infoSettings.level_edge = NOBEYOND(LEVELING_EDGE_DISTANCE_MIN, val, LEVELING_EDGE_DISTANCE_MAX);
+          infoSettings.level_edge = ((val >= edge_min)?val:edge_min);
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:109]: as for the previuous "?" and ":"

File: TFT/src/User/Menu/LevelCorner.c [hunk 32]
```diff
@@ -0,0 +1,128 @@
+#include "LevelCorner.h"
+#include "includes.h"
+
+const MENUITEMS levelcornerItems = {
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:4]: wrong indentation. cut&paste the menu at the beginning of PID.c and change the values (labels must continue to start at column 36)

**Review by digant73 — CHANGES_REQUESTED** [review]

File: TFT/src/User/Menu/LevelCorner.c [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+    {infoSettings.machine_size_min[X_AXIS] + infoSettings.level_edge, infoSettings.machine_size_min[Y_AXIS] + infoSettings.level_edge},
+    {infoSettings.machine_size_max[X_AXIS] - infoSettings.level_edge, infoSettings.machine_size_min[Y_AXIS] + infoSettings.level_edge},
+    {infoSettings.machine_size_max[X_AXIS] - infoSettings.level_edge, infoSettings.machine_size_max[Y_AXIS] - infoSettings.level_edge},
+    {infoSettings.machine_size_min[X_AXIS] + infoSettings.level_edge, infoSettings.machine_size_max[Y_AXIS] - infoSettings.level_edge}, 
+  };
+  mustStoreCmd("M401\n");
+  mustStoreCmd("G30 E0 X%d Y%d\n", (s16)pointPosition[pointer][0], (s16)pointPosition[pointer][1]);
+  mustStoreCmd("G1 Z15\n");
+  mustStoreCmd("M17 X Y Z\n");
+
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:32]: remove this blank line

File: TFT/src/User/Menu/LevelCorner.c [hunk 34]
```diff
@@ -0,0 +1,137 @@
+#include "LevelCorner.h"
+#include "includes.h"
+
+const MENUITEMS levelcornerItems = {
+  // title
+  LABEL_LEVELCORNER,
+  // icon                          label
+  {
```
Comment (digant73) [inline TFT/src/User/Menu/LevelCorner.c:8]: remove all the tabs character

**Review by digant73 — CHANGES_REQUESTED** [review]

**Comment by Guilouz** [comment]
Just try your PR and it's not working as expected for me...

When I clic on L Corner icon, probing run directly and points 2 and 4 are always in error, can do nothing on the screen and freeze after a while. This process never stops, probe of the 4 corners turns in a loop.

Here a video of this issue : https://www.youtube.com/watch?v=Miibz6StZao

With level bed corner from Marlin I don't have issue : https://www.youtube.com/watch?v=FMbsn1hwxnU

**Comment by Cal1sto** [comment]
LevelCorner.c now with the fix of the in G30 fonction modified in last Marlin bug fix release 
https://github.com/MarlinFirmware/Marlin/pull/20437
TouchMi will use the G30 E0 and other Probe like BL-Touch the G30 E1

**Comment by Cal1sto** [comment]
In preparation i will add M48 to have a visual view on the probe accuracy on this page level to help in precision decision.

**Comment by Cal1sto** [comment]
M48 added in this menu Level corner

**Comment by Cal1sto** [comment]
Due that the M48 made motor off i add the M17 call to made motor on on  X, Y and Z

**Comment by Guilouz** [comment]
PR tested and working but a small graphical issue, text overflows from icon on TFT43

162471021_2049340218546818_1918169376129697798_n

**Comment by Cal1sto** [comment]
@Guilouz Please check again on your TFT43,  should look better now , moved the text up and made only 1px shadow. Better than creating a new icon for this single function.

**Comment by Guilouz** [comment]
> @Guilouz Please check again on your TFT43, should look better now , moved the text up and made only 1px shadow. Better than creating a new icon for this single function.

There is still 1 pixel too much to the left and there the text overlaps the BLTouch icon.

**Comment by Cal1sto** [comment]
Last zn hope final generic proposition on a TFT28 should now also be good on other screen size or icon size

**Comment by oldman4U** [comment]
Maybe you want to give it a try to put the 4 blue buttons in the upper row and everything else below. Doing so it would be easier to "find" the buttons without looking too much onto the screen.

Maybe you covered this topic before: Why are the number much closed to the buttons than the text elements?

Just my 2 cents

**Comment by Guilouz** [comment]
Perfect on TFT43

6E5649CD-0799-49DD-B2B4-4983C2F3386A

**Comment by Cal1sto** [comment]
Adjusted with same procedure as the one used for displaying label, should now be same position as the one for Label

**Comment by Cal1sto** [comment]
> Maybe you want to give it a try to put the 4 blue buttons in the upper row and everything else below. Doing so it would be easier to "find" the buttons without looking too much onto the screen.
> 
> Maybe you covered this topic before: Why are the number much closed to the buttons than the text elements?
> 
> Just my 2 cents

For the 4 corners blue (cyan) icons, this disposition on the screen is spacialy the same as the bed front of the user and therefore intuitive. For this reason i will suggest to let it like that.

**Comment by Cal1sto** [comment]
Added M18 S0 X Y Z to reduce the risk when MARLIN is not set without stepper timeout, that user could not after an inactivity time move the axis, should also be integrated in Z-align procedure to reduce the move of the Z-axis after timeout is falling.

**Comment by Cal1sto** [comment]
Added after G34  (Z-Alignment) the call of M18 S0 X Y Z , to reduce the risk when Marlin is not compiled with no timeout on the stepper that the Axes could move.

**Comment by Cal1sto** [comment]
Added value refresh that was missing after setting new corner offset point.

**Comment by Cal1sto** [comment]
So nothing else to add there, is everything so far tested and ready , enjoy

**Comment by Cal1sto** [comment]
@bigtreetech So far ready

**Comment by Cal1sto** [comment]
Last fine tuning to force coordinate init on going into "L Corner", to reduce risk because the coordinateIsKnown() function is not accurate when motor are off over Marlin or other function.

**Comment by digant73** [comment]
@Cal1sto if I can suggest a layout, I would move:
* the "start" button on the left of "back" button
* the "M48" button above "back" button
* the "distance" button" on the left of the new "M48" position (so on top at 3rd position from left)
That layout will be more similar to other menus (e.g. manual leveling, MBL etc...)

**Comment by Cal1sto** [comment]
@digant73 Yes good idea it is done

**Comment by Cal1sto** [comment]
We are testing, the last change , is not validated yet

**Comment by Cal1sto** [comment]
Last menu function reordering is tested and worked, like expected

**Comment by ETE-Design** [comment]
@Cal1sto There seems to be 2 different ways to do the messuring, one mesure each point one time and give the result at the end... The other messure 1 point and set it to be the "Zero" Point, move to the next, mesure it and show the result, then you adjust it, then it messure again you can keep on messure and adjust each point before go to the next, that way you can adjust on the go to get best result... Also if using solution two would it be possible to show "live" messurement if using an inductive / capasitive sensor instead of an BLTouch...

**Comment by Cal1sto** [comment]
@ETE-Design The idea of the current implementation is to get from the mesured Z = 0 initialized with the Home (G28), the corner level , and adjust each on arround 0.01 precision, what is mostly enought . To be generic this implementation should work with all the probe able to give good accurency (see M48 function icon). So the solution provided here take the zero Z point on middle and display the difference on each corner. About inductive/capasitive sensor this should work if the accurency (precision) of the probe is enought, for this you could check it by pressing the M48 icons, what do you mean with live messurement, the live messurement is done when you press the start button.

**Comment by ETE-Design** [comment]
@Cal1sto So to understand it right it stop at each corner and let you level it before it goes to the next corner? About "Live" Messurement I was thinking about letting the sensor stand over one corner when you level it, then let the Inductive / Capasitive sensor messure all the time since it dosen't need to raise / lower the probe to messure as long as the bed is close enough it will be possible to see the exact messurement all the time... This is of cause not possible with a BLTouch since the probe need to touch the bed each time to get the messurement.

**Comment by Cal1sto** [comment]
@ETE-Design I make it generic and simple enought to all user, i do not integrate a logic analysis depending on the accurency of the probe to make this automatismus, an automatismus will be great with a motor on each screw that is yet not used on mostly printer. Therefore this version will be integrated like it is , and when you use it like i did you will be happy to have the possibility to make your bed in a better leveling state than the paper mode leveling. Try and enjoy.

**Comment by discip** [comment]
@Cal1sto
Good evening sir,
I am somewhat confused, because I can't get this to work.
I am on the latest commit https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/commit/0f464dc0e95d12f60a49a8a66fba9045e84b6274 and no matter what I do, it does not work for me.
What exactly do I have define in *Marlin* or set in *config.ini*, to make this feature work?
thank you in advance
kind regards
